osquery/tables/system/windows/users.cpp

```diff
@@ -24,81 +24,138 @@
-    SID_NAME_USE eUse;
-
-    /// MSDN guaruntees the string values returned is null terminated.
-    ret = LookupAccountSidW(
-        nullptr, sid, accntName, &accntNameLen, domName, &domNameLen, &eUse);
-    if (ret != 0) {
-      r["username"] = wstringToString(accntName);
+    WmiRequest accntReq(
+        "select Description, Name from Win32_UserAccount where sid = \"" +
+        r["uuid"] + "\"");
+    auto& accntResults = accntReq.results();
+    if (accntReq.getStatus().ok() && !accntResults.empty()) {
+      accntResults[0].GetString("Description", r["description"]);
+      accntResults[0].GetString("Name", r["username"]);
     } else {
-      TLOG << "Lookup Account by SID failed with: " << GetLastError();
+      r["description"] = "";
+      /// If there is no entry in Win32_UserAccount this is a domain user
+      wchar_t accntName[UNLEN] = {0};
+      wchar_t domName[DNLEN] = {0};
+      unsigned long accntNameLen = UNLEN;
+      unsigned long domNameLen = DNLEN;
+      SID_NAME_USE eUse;
+      ret = LookupAccountSidW(
+          nullptr, sid, accntName, &accntNameLen, domName, &domNameLen, &eUse);
+      r["username"] = ret != 0 ? wstringToString(accntName) : "";
     }
+    results.push_back(r);
+  }
 
-    /// USER_INFO_3 contains detailed info, like the uid
-    unsigned long dwDetailedUserLevel = 3;
-    LPUSER_INFO_3 pUserInfoBuffer = nullptr;
-
-    NET_API_STATUS nStatus;
-    nStatus = NetUserGetInfo(nullptr,
-                             accntName,
-                             dwDetailedUserLevel,
-                             reinterpret_cast<LPBYTE*>(&pUserInfoBuffer));
-
-    if (nStatus == NERR_Success) {
-      r["uid"] = INTEGER(pUserInfoBuffer->usri3_user_id);
-      r["uid_signed"] = INTEGER(pUserInfoBuffer->usri3_user_id);
-      r["gid"] = INTEGER(pUserInfoBuffer->usri3_primary_group_id);
-      r["gid_signed"] = INTEGER(pUserInfoBuffer->usri3_primary_group_id);
-      r["description"] =
-          SQL_TEXT(wstringToString(pUserInfoBuffer->usri3_comment));
-    } else {
-      /// If NetUserGetInfo fails parse the uid manually
-      auto sidTokens = osquery::split(sidString, "-");
-      if (sidTokens.size() == 8) {
-        r["uid"] = INTEGER(sidTokens.back());
-        r["uid_signed"] = INTEGER(sidTokens.back());
+  /// Lastly do a sweep for any accounts that don't yet have a profile
+  unsigned long dwUserInfoLevel = 3;
+  unsigned long dwNumUsersRead = 0;
+  unsigned long dwTotalUsers = 0;
+  unsigned long resumeHandle = 0;
+  LPBYTE userBuffer = nullptr;
+  do {
+    ret = NetUserEnum(nullptr,
+                      dwUserInfoLevel,
+                      0,
+                      &userBuffer,
+                      MAX_PREFERRED_LENGTH,
+                      &dwNumUsersRead,
+                      &dwTotalUsers,
+                      &resumeHandle);
+
+    if ((ret == NERR_Success || ret == ERROR_MORE_DATA) &&
+        userBuffer != nullptr) {
+      auto iterBuff = LPUSER_INFO_3(userBuffer);
```

Comment:
Would it make sense to place this "conditional-pass, then loop" logic into another method? Or are there too many variables you'd have to pass into that new scope?